### DISCUSSION TITLE
Would you prefer if React CSS modules enables Code lens and Peek References features

### DISCUSSION THREAD

**Discussion by Viijay-Kr** [Viijay-Kr/react-ts-css#133] [Q&A] [answered]
Dear Community members , 

I have been thinking of introducing of some CSS language features namely Code lens and Peek references for selectors 
 * [Reference Information](https://code.visualstudio.com/docs/editor/editingevolved#_reference-information)
   * [Demo](https://github.com/Viijay-Kr/react-ts-css/blob/main/assets/references.gif)
 * [Code lens](https://code.visualstudio.com/api/language-extensions/programmatic-language-features#codelens-show-actionable-context-information-within-source-code)
   * [Demo](https://github.com/Viijay-Kr/react-ts-css/blob/main/assets/code-lens.gif)

The above two features has been prototyped and being experimented in this [PR](https://github.com/Viijay-Kr/react-ts-css/pull/126)
   
I have definitely experienced the need for the above features when working with large react projects that involves a huge deal of selectors which are modular . A quick action to find the usages of the selectors is super helpful to perform any further refactoring.

However the current prototype I have drafted comes with a performance hiccup both run time and memory. I have tried to optimize it as much as possible but still not ready to ship it without optimizing further and provide a zero cost solution. 
In order to do that, I first need your inputs and suggestions to proceed with those features.  

So answering the questions will be helpful to proceed any further 
1. Do you see the any value from these features ?
2. Do you absolutely need them or can live without them ? 

If you have come across this thread on your spare time , do throw in your feedback. It will be helpful for me to take a decision and spend further time on it 
cc: @karlhorky @strlns

**Comment by karlhorky** [comment, marked as answer]
These would be great! 😍

Personally, I would gladly pay a small performance penalty for the Peek Reference feature. If it slowed down my editor too much, I would disable it (hopefully there would be a way to disable)

I would personally disable the CodeLens version, since I disable all CodeLens decorations in my editor - I like having a consistent line height.

**Reply by Viijay-Kr** [reply]
@karlhorky Thanks for taking the time to respond. I agree with you about code lens . It definitely gives a bothering experience at times .

I will try to ship Reference Feature soon since you have showed interest in that . I could only hope that it brings the same value to our users :)

**Comment by Viijay-Kr** [comment]
References and Code lens feature is out with v2.6.0

**Reply by karlhorky** [reply]
Thanks for this!

> Personally, I would gladly pay a small performance penalty for the Peek Reference feature. 
> 
> I would personally disable the CodeLens version

I finally got some time to upgrade to [`viijay-kr.react-ts-css@3.1.1`](https://marketplace.visualstudio.com/items?itemName=viijay-kr.react-ts-css) and I'm trying to achieve the feature set I mentioned above: 

- Peek References on
- CodeLens off

I haven't yet figured out how to activate the Peek Reference feature without the CodeLens - is this possible?

In JavaScript / TypeScript files I usually <kbd>cmd</kbd>-click on variables and functions to show the peek references, but this doesn't seem to do anything here - maybe it would be possible to add this shortcut to show the peek references in the CSS files as well?

**Reply by karlhorky** [reply]
Oh, I found two ways to do it (these actions can also be assigned to keyboard shortcuts):

1. <kbd>cmd</kbd>-<kbd>shift</kbd>-<kbd>P</kbd> -> Peek References
2. Right click -> Peek -> Peek References (or right click -> Go to References)

Would be cool to get the <kbd>cmd</kbd>-click option as well, if possible!

**Reply by karlhorky** [reply]
Interesting, maybe this <kbd>cmd</kbd>-click is actually "Go to Definition" (more info below), which currently doesn't work as an action:

![Screenshot 2024-04-06 at 18 24 54](https://github.com/Viijay-Kr/react-ts-css/assets/1935696/bb9cdd93-8d37-4125-902a-b2465a73760f)

So maybe if a definition is contributed by the extension, then the <kbd>cmd</kbd>-click would also work 🤔

Reference about "Go to Definition":

> The Go To Definition and Open Link mouse gestures will adapt such that they do not conflict with the multicursor modifier.

Source: https://stackoverflow.com/a/57446936/1268612

**Reply by Viijay-Kr** [reply]
@karlhorky Actually I'm not quite sure about how to accomplish this using language providers ? My bet is actually registering a definition provider for css language and get the references for the range. Feel free to open a issue if you can


Also I sorta agree with you that cmd+click would be greatly useful than doing Peek references.

**Reply by Viijay-Kr** [reply]
@karlhorky After digging a bit about how typescript documents support this feature . Turns out you have an option to assign an alternate command for go to definition . This is set as peek references in my settings.json so when a symbol doesn't yield any go to definition , the peek references command is executed . Unfortunately this doesn't apply/doesn't work in css documents. So I presume this is an issue with vs code rather than this extension as VS code doesn't have builtin cross language support. 

Edit: Although Go to definition works when the cursor is at the beginning of a selector


https://github.com/Viijay-Kr/react-ts-css/assets/8697234/a1c862dd-088e-4b6e-a00d-b660633fc004

So it works partially but not completely . Registering go to definition is a dirty workaround

**Reply by karlhorky** [reply]
@Viijay-Kr thanks for investigating! Do you think this issue is relevant?

- https://github.com/microsoft/vscode/issues/47331

If it is, does it make sense for you to explain the use case over there in case the VS Code team would be willing to consider it?
